### ISSUE TITLE
seafile-server startup hangs and restarts at sqlite3_exec()

### ISSUE THREAD

**Issue by T-X** [haiwen/seafile-server#129]
I'm currently trying to run seafile/seahub v6.2.5 via "seaf-admin start" as a non-root user from a local installation (all libraries and headers installed to ~/install/).

Unfortunately, seafile-server never seems to fully start, I never see a socket for port 8082. I also noticed that seafile-server was restarted frequently (the pid file number changed about every ten seconds).

I added some more debug output to seafile-server and it always hangs when seafile-server tries to create this table:

``CREATE TABLE IF NOT EXISTS SystemInfo (info_key VARCHAR(256), info_value VARCHAR(1024))``

More specifically, it hangs with the sqlite3_exec() call via sqlite3_blocking_exec() (via sqlite_db_connection_execute() via db_connection_execute() via schedule_create_system_default_repo().

Even though there is a SQLITE_OPEN_SHAREDCACHE for sqlite3_open_v2() in sqlite_get_db_connection(), I tried adding an explicit sqlite3_enable_shared_cache(1) before the sqlite3_open_v2() call. It returned SQLITE_OK, but made no difference.

In the sqlite3 documentation [here](https://sqlite.org/unlock_notify.html) I noticed that there were only examples for sqlite3_prepare_v2() and sqlite3_step(). Could it be, that sqlite3_exec() does not work like that or never returns SQLITE_LOCKED?

**Comment by T-X** [comment]
Okay, further debugging revealed it's actually not sqlite3 hanging. Before the creation of the table mentioned above, a few lines earlier, a thread for http_server_run is started. And this thread crashes at "access_file_init() => evhtp_set_regex_cb(htp, "^/blks/.*", access_blks_cb, NULL);" in server/access-file.c:

https://gist.github.com/T-X/df35ddf44a19a356d017cce9fffd969e

So it seems just a coincidence that the http-server thread always crashes while the main thread is executing this particular sqlite3 statement.

I'm using libevhtp0 / libevhtp-dev version 1.2.11-3 as provided by Debian Stretch (current Debian stable).

**Comment by niol** [comment]
seafile-server requires libevhtp >= 1.2.13 . Otherwise it hangs. It also requires some patches to seafile-server . See #67 for further details.

**Comment by T-X** [comment]
And here with valgrind (and now libevhtp0-dbg installed, too - unfortunately the line numbers for libevhtp still seem to make no sense):

https://gist.github.com/T-X/ec13091742f832a5def3ee838e342dd9

**Comment by T-X** [comment]
@niol: Ah, thanks a lot for your quick reply! I wiill have a look at and will report back later. And will probably file a bug report to Debian then if an outdated version of libevhtp (and other missing bugfixes) are the issue.

**Comment by T-X** [comment]
@niol: Here's the report:

tl;dr: It works :-)! I'm now running seafile 6.2.5 with libevhtp 1.2.16. The crashes and complaints from valgrind mentioned earlier are gone. So the main issue indeed was the nasty regex_t size mismatch...

To get this recent version of libevhtp running I followed the changes you mentioned in #67 in [here](https://github.com/haiwen/seafile-server/issues/67#issuecomment-337904800) (excuding the second bullet, the fix for criticalstack/libevhtp#38 which is already part of libevhtp 1.2.16).

---

However, while playing will valgrind I experienced the following two issues. Neither would happen if running without valgrind, so they seem to be race conditions triggered by the slow-down through valgrind.

**Null pointer exception in epoll_wait() via event_base_loop()**

When starting seaf-server with valgrind, the first startup would crash in http_server_run()/ccnet_main() -> event_base_loop() -> ... -> epoll_wait() with a null pointer exception:

<pre><code>[...]
==13591== Thread 2:
==13591== Syscall param epoll_pwait(sigmask) points to unaddressable byte(s)
==13591==    at 0x56C4074: epoll_pwait (epoll_pwait.c:42)
==13591==    by 0x496D4E7: epoll_dispatch (epoll.c:407)
==13591==    by 0x4956E4F: event_base_loop (event.c:1607)
==13591==    by 0x135777: http_server_run (http-server.c:2141)
==13591==    by 0x54B609F: start_thread (pthread_create.c:335)
==13591==    by 0x56C3EDB: thread_start (clone.S:77)
==13591==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
[...]
</code></pre>

https://gist.github.com/T-X/e85eb8a919d05da2c859f81a515c1e8c#file-seaf-server-valgrind-13581-1518819171-log

Interestingly, a subsequent restart from the seafile-controller would not yield this crash anymore. So probably some caching speed-up hides this race condition again.

**"bind fails: Address already in use" / Failed to start listen manager.**

After the initial null pointer exception mentioned above all restarts tried by the seafile controller failed for me with this bind error:

<pre><code>[...]
[02/16/2018 10:13:19 PM] net.c(193): bind fails: Address already in use
[02/16/2018 10:13:19 PM] listen-mgr.c(98): [listen mgr] failed to bind port 12001
[02/16/2018 10:13:19 PM] seafile-session.c(242): Failed to start listen manager.
[...]
</code></pre>

https://gist.github.com/T-X/e85eb8a919d05da2c859f81a515c1e8c#file-seafile-log

Either running without valgrind or by increasing the controller interval (I changed it from 10s to 75s) would make this issue go away.

My guess is that the mix of a watchdog behavior implemented through the seafile controller and an unclean program exit via exit(), so without closing any file or socket descriptors, is probably a bad idea. I would not be surprised if the Linux kernel (I'm testing with an aarch64 Linux 4.14.12-v8+ kernel as provided by the Raspberry Pi project, running on a Pi 3) needs a few seconds to recognize that the process assigned to a descriptor is gone when the descriptors are not closed explicitly.

It would probably be wise to always try to close descriptors, maybe via an atexit() handler. Or maybe better replace all exit() calls with a proper error and shutdown handling in seaf-server (and libevhtp, too). It could be wise to implement a back-off interval into the seafile-controller, too. If restarts are triggered subsequently, then for instance double the check period every time up to a certain maximum. Or better, let just systemd do all this watchdog behavior, I think it already has such a back-off feature.

---

Full logs of a single "seaf-admin start" run with valgrind where both issues occurred:

https://gist.github.com/T-X/e85eb8a919d05da2c859f81a515c1e8c

**Comment by niol** [comment]
Glad it works.
For your other issues, I cannot help as I completely skipped using ```seafile-controller``` and replaced it with [a systemd unit](https://sml.zincube.net/~niol/repositories.git/seafile-server/plain/debian/seafile-server.service).

**Comment by DevelAngel** [comment]
This issue is open over more than a year. But anyway, I will add my observations with newer releases because I think it is related to this issue. My issue is reproducable every startup of seafile server.

Seafile-Server Environment:
- OS: Gentoo Linux (arm64)
- Hardware: Raspberry Pi 3b
- seafile-server: 7.0.5
- ccnet-server: 7.0.5
- libevhtp: 1.2.18 (from criticalstack, seafile-server slightly patched)
- libevent: 2.1.8
- libsearpc: 3.1.0
- glib: 2.60.6

> Okay, further debugging revealed it's actually not sqlite3 hanging. Before the creation of the table mentioned above, a few lines earlier, a thread for http_server_run is started. And this thread crashes at "access_file_init() => evhtp_set_regex_cb(htp, "^/blks/.*", access_blks_cb, NULL);" in server/access-file.c. Full gdb output of the crashing thread:
> https://gist.github.com/T-X/df35ddf44a19a356d017cce9fffd969e

In release 7.0.5, I have similar issues (with similar backtrace): In the `upload_file_init()` function in file `server/upload-file.c` the following line produces SIGABRT signal inside `regcomp()` function:
```c++
cb = evhtp_set_regex_cb (htp, "^/upload-raw-blks-api/.*", upload_raw_blks_api_cb, NULL);
```

The GDB says the following:
```gdb
Starting program: seaf-server -F /var/lib/seafile/data/conf -c /var/lib/seafile/data/ccnet -d /$
ar/lib/seafile/data/seafile-data -l /var/lib/seafile/data/logs/seafile.log -P /pids/seaf-server.pid -f                                 
[Thread debugging using libthread_db enabled]                                                                                          
Using host libthread_db library "/lib64/libthread_db.so.1".                                                                            
[New Thread 0x7fbda1e160 (LWP 24708)]                                                                                                  
[New Thread 0x7fbd21d160 (LWP 24709)]                                                                                                  
[New Thread 0x7fbca1c160 (LWP 24710)]                                                                                                  
mremap_chunk(): invalid pointer

Thread 3 "seaf-server" received signal SIGABRT, Aborted.                                                                               
[Switching to Thread 0x7fbd21d160 (LWP 24709)]                                                                                         
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50                                                                  
50        return ret;                                                                                                                  
(gdb) bt                                                                                                                               
#0  0x0000007fbe66929c in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50       
#1  0x0000007fbe6557e8 in __GI_abort () at abort.c:79                                                                                  
#2  0x0000007fbe6a7afc in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7fbe773540 "%s\n")                             
    at ../sysdeps/posix/libc_fatal.c:181                                                                                               
#3  0x0000007fbe6afda0 in malloc_printerr (str=str@entry=0x7fbe76ec48 "mremap_chunk(): invalid pointer") at malloc.c:5352              
#4  0x0000007fbe6b02ec in mremap_chunk (p=p@entry=0x7fb800bca0, new_size=new_size@entry=208) at malloc.c:2868                          
#5  0x0000007fbe6b4f10 in __GI___libc_realloc (oldmem=0x7fb800bcb0, bytes=200) at malloc.c:3209                                        
#6  0x0000007fbe6eb1fc in re_dfa_add_node (dfa=dfa@entry=0x7fb8028410, token=...) at regex_internal.c:1431
#7  0x0000007fbe6eb2e8 in duplicate_node (dfa=dfa@entry=0x7fb8028410, org_idx=org_idx@entry=1, constraint=constraint@entry=16)
    at regcomp.c:1614                                              
#8  0x0000007fbe6eb538 in duplicate_node_closure                                                                                       
    (dfa=dfa@entry=0x7fb8028410, top_org_node=top_org_node@entry=0, top_clone_node=top_clone_node@entry=0, root_node=root_node@entry=0$
 init_constraint=<optimized out>) at regcomp.c:1537                                                                                    
#9  0x0000007fbe6ef500 in calc_eclosure_iter
    (dfa=dfa@entry=0x7fb8028410, node=node@entry=0, root=root@entry=true, new_set=<optimized out>) at regcomp.c:1718
#10 0x0000007fbe6f75e8 in calc_eclosure (dfa=<optimized out>) at regcomp.c:1681
#11 0x0000007fbe6f75e8 in analyze (preg=0x7fb8027ea0) at regcomp.c:1204
#12 0x0000007fbe6f75e8 in re_compile_internal
    (preg=preg@entry=0x7fb8027ea0, pattern=pattern@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", length=<optimized out>, syntax=synta$
@entry=16843462) at regcomp.c:795
#13 0x0000007fbe6f8574 in __GI___regcomp
    (preg=0x7fb8027ea0, pattern=pattern@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", cflags=cflags@entry=16) at regcomp.c:491
#14 0x0000007fbf684e20 in evhtp_callback_new
    (path=path@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", type=type@entry=evhtp_callback_type_regex, cb=0x555558a768 <upload_raw_b$
ks_api_cb>, arg=0x0) at /usr/src/debug/net-libs/libevhtp-1.2.18-r1/libevhtp-1.2.18/evhtp.c:4224
#15 0x0000007fbf6851c8 in evhtp_set_regex_cb
    (htp=htp@entry=0x7fb8001050, pattern=pattern@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", cb=cb@entry=0x555558a768 <upload_raw_bl
ks_api_cb>, arg=arg@entry=0x0) at /usr/src/debug/net-libs/libevhtp-1.2.18-r1/libevhtp-1.2.18/evhtp.c:4649
#16 0x000000555558b828 in upload_file_init (htp=0x7fb8001050, http_temp_dir=<optimized out>) at upload-file.c:2913
#17 0x00000055555834f0 in http_request_init (server=0x5555659c10, server=0x5555659c10) at http-server.c:2122
#18 0x00000055555834f0 in http_server_run (arg=0x5555659c10) at http-server.c:2230
#19 0x0000007fbec56e6c in start_thread (arg=0x7fffffec8e) at pthread_create.c:486
#20 0x0000007fbe712ffc in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:78
```

After I commented the code line out, the server starts up.
A colleage mentioned that alignment problems could raise such weird behaviors (when the pointer will be deferenced).

**Comment by DevelAngel** [comment]
> For your other issues, I cannot help as I completely skipped using `seafile-controller` and replaced it with [a systemd unit](https://sml.zincube.net/~niol/repositories.git/seafile-server/plain/debian/seafile-server.service).

I will do the same for my seafile-server instance. The seafile controller is a good way to test the server because it starts also the ccnet server. But the controller seems not to integrate well into init systems and its requirements for services. I prefer also systemd service files.

**Comment by niol** [comment]
> #13 0x0000007fbe6f8574 in __GI___regcomp
>     (preg=0x7fb8027ea0, pattern=pattern@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", cflags=cflags@entry=16) at regcomp.c:491
> #14 0x0000007fbf684e20 in evhtp_callback_new
>     (path=path@entry=0x55555dfb88 "^/upload-raw-blks-api/.*", type=type@entry=evhtp_callback_type_regex, cb=0x555558a768 <upload_raw_b$

> After I commented the code line out, the server starts up.
> A colleage mentioned that alignment problems could raise such weird behaviors (when the pointer will be deferenced).

This reminds me of the problem with libonigurama linking to which I proposed a fix : https://github.com/haiwen/seafile-server/pull/12#discussion_r137513336

If this is the same thing, it makes me sad that you still wasted some more time on this...
